src/wallet/wallet2.cpp

```diff
@@ -6393,56 +6393,99 @@ void wallet2::transfer_selected_rct(std::vector<cryptonote::tx_destination_entry
 
+	// Highest and second highest available amounts
+	uint64_t amount_a = 0;
+	uint64_t amount_b = 0;
+	// Check if the pick_list is sorted (sorting a sorted list is a worst case scenario usually)
+	uint64_t last_block = 0;
+	bool sorted = true;
+
 	// try to find a rct input of enough size
 	for(size_t i = 0; i < m_transfers.size(); ++i)
 	{
 		const transfer_details &td = m_transfers[i];
-		if(!td.m_spent && td.is_rct() && td.amount() >= needed_money && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
+		if(!td.m_spent && !td.m_key_image_partial && td.is_rct() && is_transfer_unlocked(td) && 
+				td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
 		{
-			LOG_PRINT_L2("We can use " << i << " alone: " << print_money(td.amount()));
-			picks.push_back(i);
-			return picks;
+			uint64_t amt = td.amount();
+			if(amt >= needed_money)
+			{
+				LOG_PRINT_L2("We can use " << i << " alone: " << print_money(amt));
+				picks.push_back(i);
+				return picks;
+			}
+
+			pick_list.emplace_back(i, amt, td.m_block_height);
+			if(amt > amount_a)
+			{
+				amount_b = amount_a;
+				amount_a = amt;
+			}
+
+			if(td.m_block_height < last_block)
+				sorted = false;
+			last_block = td.m_block_height;
 		}
 	}
 
+	// This means there is no chance of constructing a two output tx of any kind, return empty
+	if(amount_a + amount_b < needed_money)
+		return picks;
+
+	// This makes the O(n^2) below the worst possible case and limited to small n's (otherwise they will naturally spead out)
+	if(!sorted)
+		std::sort(pick_list.begin(), pick_list.end(), [](const pick_out& a, const pick_out& b) { return a.blk_height < b.blk_height; });
+	else
+		LOG_PRINT_L2("pick_preferred_rct_inputs: sort skipped, we are sorted already");
+
 	// then try to find two outputs
 	// this could be made better by picking one of the outputs to be a small one, since those
 	// are less useful since often below the needed money, so if one can be used in a pair,
 	// it gets rid of it for the future
-	for(size_t i = 0; i < m_transfers.size(); ++i)
-	{
-		const transfer_details &td = m_transfers[i];
-		if(!td.m_spent && !td.m_key_image_partial && td.is_rct() && is_transfer_unlocked(td) && td.m_subaddr_index.major == subaddr_account && subaddr_indices.count(td.m_subaddr_index.minor) == 1)
-		{
-			LOG_PRINT_L2("Considering input " << i << ", " << print_money(td.amount()));
-			for(size_t j = i + 1; j < m_transfers.size(); ++j)
-			{
-				const transfer_details &td2 = m_transfers[j];
-				if(!td2.m_spent && !td.m_key_image_partial && td2.is_rct() && td.amount() + td2.amount() >= needed_money && is_transfer_unlocked(td2) && td2.m_subaddr_index == td.m_subaddr_index)
+	for(size_t i = 0; i < pick_list.size(); i++)
+	{
+		LOG_PRINT_L2("Considering input " << pick_list[i].idx << ", " << print_money(pick_list[i].amount));
+		for(size_t j = pick_list.size(); j-- > i+1;)
```

Comment:
now it is a out of memory acces because j is in the first round one element behind the last element in the container.

Reply:
All fine j is decremented after the comparison